ai_flow/task_executor/kubernetes/k8s_task_executor.py

```diff
@@ -0,0 +1,127 @@
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+import logging
+import multiprocessing
+import threading
+from multiprocessing.managers import SyncManager
+from typing import Optional
+from queue import Queue
+
+from kubernetes import client
+from kubernetes.client import V1Pod
+from kubernetes.client.rest import ApiException
+
+from ai_flow.common.exception.exceptions import AIFlowException
+from ai_flow.model.task_execution import TaskExecutionKey
+from ai_flow.task_executor.task_executor import BaseTaskExecutor
+from . import helpers
+from .helpers import make_safe_label_value, POISON, get_kube_client
+from .kube_config import KubeConfig
+from .kube_scheduler import KubernetesScheduler
+from .kube_watcher import KubernetesJobWatcher, ResourceVersion
+from ai_flow.common.configuration import config_constants
+from ai_flow.common.util.thread_utils import StoppableThread
+
+logger = logging.getLogger(__name__)
+
+
+class KubernetesTaskExecutor(BaseTaskExecutor):
+    def __init__(self):
+        self.manager: Optional[SyncManager] = None
+        self.task_queue: Optional['Queue[TaskExecutionKey]'] = None
+        self.kube_scheduler: Optional[KubernetesScheduler] = None
+        self.kube_watcher: Optional[KubernetesJobWatcher] = None
+        self.kube_config = KubeConfig(config_constants.K8S_TASK_EXECUTOR_CONFIG)
+        self.kube_client = helpers.get_kube_client(in_cluster=self.kube_config.is_in_cluster(),
+                                                   config_file=self.kube_config.get_config_file())
+        self._process_observer = StoppableThread(target=self._check_watcher_alive)
+
+    def start_task_execution(self, key: TaskExecutionKey):
+        if not self.task_queue:
+            raise AIFlowException('KubernetesTaskExecutor not started.')
+        self.task_queue.put(key)
+
+    def stop_task_execution(self, key: TaskExecutionKey):
+        dict_string = "workflow_execution_id={},task_name={},seq_number={}".format(
+            make_safe_label_value(key.workflow_execution_id),
+            make_safe_label_value(key.task_name),
+            make_safe_label_value(key.seq_num),
+        )
+        kwargs = dict(label_selector=dict_string)
+        pod_list = self.kube_client.list_namespaced_pod(namespace=self.kube_config.get_namespace(), **kwargs)
```

Comment:
Why not stop the pod by id

Reply:
Currently, we don't find listing causes performance issues, we can try another interface in the future if we find out the listing is inefficient.